### ISSUE TITLE
Create an appropriate folder structure in the repository

### ISSUE THREAD

**Issue by ash-teach** [nic-dgl104-winter-2024/pattern-library#1]
To be included:

- Design patterns with implementations in specific languages
- Software architecture patterns with implementations/descriptions in specific languages
- Open source [documents](https://opensource.guide/starting-a-project/)

**Comment by ash-teach** [comment]
Maybe check out the structures found in similar libraries on GitHub and elsewhere?

**Comment by ting-dev-coder** [comment]
I found some similar repositories 
[Design-Patterns-In-Kotlin](https://github.com/dbacinski/Design-Patterns-In-Kotlin)
[javascript-patterns](https://github.com/chuanxshi/javascript-patterns)
[awesome-design-patterns](https://github.com/DovAmir/awesome-design-patterns)
[java-design-patterns](https://github.com/iluwatar/java-design-patterns)

Perhaps starting with a patterns-first approach and then diving into specific languages?

> pattern-library
|-- docs/patterns
|   |-- observer
|   |   |-- javascript
|   |   |   |-- ....
|   |-- singleton
|   |   |-- javascript
|   |   |   |-- ....
|-- assets
|   |-- .....
|-- README.md

I have seen some people categorize patterns into different behaviors like (Structural, Behavioral, Creational), but I feel like it is too abstract for beginners to start with.

**Comment by ash-teach** [comment]
Yup, I agree. No need to add the extra layer of complexity (e.g. Creational, Behavioural, etc.). 

@ting-dev-coder - Is the structure you intended more like this:

```
pattern-library
|-- docs/patterns
| |-- observer
| | |-- javascript
| | |-- ....
| |-- singleton
| | |-- javascript
| | |-- ....
|-- assets
| |-- .....
|-- README.md 
```
I assume the `...` is mean to represent additional implementations? (i.e. in Kotlin, Dart, Lua, etc.), in which case it should be at the same level as the JavaScript folder.

Also, I wonder if we need the `patterns` folder to be under a `docs` folder? They may not really be "documents" afterall.

**Comment by ting-dev-coder** [comment]
The use of `...` represents etcetera, I hadn't come up with some possible files, so I put `...`, but never mind, they should be at the same level; I mistakenly placed them incorrectly.

@ash-teach Do you think we need individual example code files, each with their respective language types, even though they might be included in README.md?
for example :
```
| | |-- javascript
| | | |-- README.md
| | | |-- singleton.js
| | | |-- advance-singleton.js

```

As for the folder, you are right `patterns` folder might not be documents.
Sorry for any confusion. I struggled to find a suitable root file name between 'docs' and 'patterns,' so so I used 'splash'  to mean **or**.

**Comment by ash-teach** [comment]
Ok. That looks good. As for the individual README.md docs: let's leave that off the table for now, but we can come back to it, if it appears that the documentation is ballooning in size.

**Comment by KaiKitKobold** [comment]
If this is still open and @ting-dev-coder, if you're not implementing this at the current moment, I can add the folder Roots for each of these languages (Besides JavaScript)

**Comment by ting-dev-coder** [comment]
I was not assuming the code and document are separated, and though document for each type of language.
After finishing this commit #2, I came out with another structure
```
| | |-- singleton
| | | |-- singleton.js
| | | |-- singleton-advance.js
| | | |-- singleton.kts
| | | |-- singleton.dart
| | | |-- README.md
```
I think we also need some file name conventions, for example, start with the pattern and connect words with -(Hyphen),  so they can easily listed together.
Let me know if you have any advice or other ideas.
@KaiKitKobold @ash-teach

**Comment by KaiKitKobold** [comment]
I have just created a quick folder commit using @ash-teach's pattern across each language so students can work on code right away instead of having to make separate nested implementations. 

```
pattern-library
|-- docs/patterns
| |-- observer
| | |-- javascript
| | |-- kotlin
| | |-- dart
| | |-- react_native 
| | |-- lua 
| |-- singleton
| | |-- javascript
| | |-- dart
| | |-- react_native 
| | |-- lua 
|-- assets
| |-- .....
|-- README.md 

```

This is my pull request #32 
This is the following pull request I've initially committed to; we can try to add further complexity to the bat,
Please tell me If I should add more complexity before this is pulled to the main. 

I'll update this comment with more discussion once I look at some research regarding what we should do for our file name conventions since that will be important near the start of this project.

**Comment by KaiKitKobold** [comment]
So, I accidentally committed in the wrong place, which caused the issue; I'll attempt to update the fork to provide the proper folder structure. 

Regarding how we should categorize, I do think we should take a page from other repositories, such as the one that Ting Yu researched, or a repository such as [This dart design pattern library](https://github.com/IliaKhuzhakhmetov/DesignPatterns-Dart)

I do think we should add categories for the creational patterns, structural patterns, and behavioral patterns, while this may initially confuse some students, It is an important skill to categorize and identify each of them, and can help with organization, To help with this, I can update the read. me to provide hyperlinks to each of the nested patterns or describe it as a list 

Overall this can be useful for organization especially if we add additional patterns to the repository, what do you think @ash-teach?

**Comment by lcgraves** [comment]
Hi @ash-teach, @KaiKitKobold and @ting-dev-coder, I like the approach of having different implementations of a design pattern in different documents. I'm wondering if it would be a good idea to give them more descriptive file names to make it clearer what our approach is and how we're organizing our library. 

So rather than just singleton.js and singleton-advance.js we could have something like:

singleton_basic.js
singleton_thread_safe.js
singleton_lazy_initialization.js

Thoughts?

**Comment by KaiKitKobold** [comment]
> Hi @ash-teach, @KaiKitKobold and @ting-dev-coder, I like the approach of having different implementations of a design pattern in different documents. I'm wondering if it would be a good idea to give them more descriptive file names to make it clearer what our approach is and how we're organizing our library.
> 
> So rather than just singleton.js and singleton-advance.js we could have something like:
> 
> singleton_basic.js singleton_thread_safe.js singleton_lazy_initialization.js
> 
> Thoughts?

That could potentially work, but I think that's more of a naming convention for the pattern library issue commit, rather than the current issue at the moment, but I think that's an smart idea for naming conventions.

**Comment by KaiKitKobold** [comment]
By the way, I've done another pull request @ash-teach this time. It should work entirely as it depends on it; let me know if I need to add more or remove certain aspects. 

Pull Request: #42

**Comment by ash-teach** [comment]
Closed by #42

**Comment by KaiKitKobold** [comment]
Hi, I've updated this to include the new content the recent pull requests created.
The pull request is #82
